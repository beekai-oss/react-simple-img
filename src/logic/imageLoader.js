// @flow
import filterImgSrc from '../utils/filterSrcset';
import fetchImage from './fetchImage';
import applyImage from './applyImage';
import logError from '../utils/logError';
import setImageHeight from '../utils/setImageHeight';
import updateSessionStorage from './updateSessionStorage';

export default function imageLoader(target: any) {
  try {
    const image = new Image(); // eslint-disable-line no-undef

    const { observer, imgLoadingRefs } = window.__REACT_SIMPLE_IMG__;
    const src = filterImgSrc(target);

    observer.unobserve(target);
    imgLoadingRefs.set(target, image);

    if (!src) {
      logError('Filter Image source returned empty image source', target);
      return;
    }

    if (target.parentNode && target.parentNode.style.height === '1px') {
      setImageHeight(image, target);
    }

    fetchImage(image, src)
      .then(() => {
        if (target) {
          applyImage(target, image, src);

          if (!window.__REACT_SIMPLE_IMG__) return;
          const { disableAnimateCachedImg, callBackRefs } = window.__REACT_SIMPLE_IMG__;

          if (disableAnimateCachedImg) {
            updateSessionStorage(src);
          }

          const callback = callBackRefs.get(target);

          if (callback) {
            callback();
            callBackRefs.delete(target);
          }
        }
      })
      .catch(e => {
        logError('Fetch image failed with target', target, e);
      });
  } catch (e) {
    logError('Image loader failed with target', target, e);
  }
}
