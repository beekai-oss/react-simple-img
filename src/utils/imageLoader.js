// @flow
import filterImgSrc from '../logic/filterSrcset';
import fetchImage from './fetchImage';
import applyImage from './applyImage';
import logError from './logError';

export default function imageLoader(target: any) {
  try {
    const image = new Image(); // eslint-disable-line no-undef

    if (this) {
      this.observer.unobserve(target);
      this.appendImgLoadingRef(image);
    } else {
      const { observer, imgLoadingRefs } = window.__REACT_SIMPLE_IMG__;

      observer.unobserve(target);
      imgLoadingRefs.set(target, image);
    }

    const src = filterImgSrc(target);

    if (!src) {
      logError('Filter Image source returned empty image source', target);
      return;
    }

    fetchImage(image, src)
      .then(() => {
        applyImage.apply(this, [target, image, src]);
      })
      .catch(e => {
        logError('Fetch image failed with target', target, e);
      });
  } catch (e) {
    logError('Image loader failed with target', target, e);
  }
}
