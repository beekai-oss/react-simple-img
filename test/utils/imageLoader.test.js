import imageLoader, { applyImage } from '../../src/utils/imageLoader';
jest.mock('../../src/utils/fetchImage', () => () => new Promise(resolve => resolve('test')));

describe('imageLoader', () => {
  describe('when preloadImage called', () => {
    it('should catch error failed', async () => {
      try {
        await imageLoader({});
      } catch (e) {
        expect(e).toMatchSnapshot();
      }
    });

    it('should fetch the image and apply src to original image', async () => {
      const unobserveSpy = jest.fn();
      const setStateSpy = jest.fn();
      const target = { dataset: { src: 'test' } };
      await imageLoader.call(
        {
          observer: {
            unobserve: unobserveSpy,
          },
          setState: setStateSpy,
        },
        target,
      );

      expect(unobserveSpy).toHaveBeenCalled();
      expect(setStateSpy).toHaveBeenCalled();
    });
  });

  it('should update mount images state when applyImage is called', () => {
    const setAttributeSpy = jest.fn();
    const getAttributeSpy = jest.fn();
    expect(
      applyImage(
        {
          src: 'test',
          style: {
            visibility: 'hidden',
          },
          nextSibling: {
            setAttribute: setAttributeSpy,
            getAttribute: getAttributeSpy,
          },
        },
        'loadedImg',
      ),
    ).toMatchSnapshot();
    expect(getAttributeSpy).toHaveBeenCalled();
    expect(setAttributeSpy).toHaveBeenCalled();
  });
});
