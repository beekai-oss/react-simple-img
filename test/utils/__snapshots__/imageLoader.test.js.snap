// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`imageLoader should update mount images state when applyImage is called 1`] = `undefined`;

exports[`imageLoader when preloadImage called should catch error failed 1`] = `[Error: ðŸ’© Fetch image failed with target {} and error message TypeError: Cannot read property 'unobserve' of undefined]`;
